### PR TITLE
Replication performance tests.

### PR DESCRIPTION
**Pull request by dkorolev**
Interestingly, degrades with log size.

T420s:

```
N = 100, M = 1
Memory
Iteration 1/10, 6496.062992 replications per second.
Iteration 2/10, 7327.905255 replications per second.
Iteration 3/10, 6543.723974 replications per second.
Iteration 4/10, 6655.014789 replications per second.
Iteration 5/10, 5969.249322 replications per second.
Iteration 6/10, 6344.527044 replications per second.
Iteration 7/10, 5265.397298 replications per second.
Iteration 8/10, 6204.562547 replications per second.
Iteration 9/10, 6582.009175 replications per second.
Iteration 10/10, 7354.033576 replications per second.
Replications per second: 6474.248597 (stddev 579.503816).
File
Iteration 1/10, 4713.163532 replications per second.
Iteration 2/10, 5230.070263 replications per second.
Iteration 3/10, 7259.128905 replications per second.
Iteration 4/10, 5325.443787 replications per second.
Iteration 5/10, 5279.718415 replications per second.
Iteration 6/10, 4359.121131 replications per second.
Iteration 7/10, 6217.421340 replications per second.
Iteration 8/10, 7451.452657 replications per second.
Iteration 9/10, 5582.496899 replications per second.
Iteration 10/10, 3626.506465 replications per second.
Replications per second: 5504.452339 (stddev 1140.462758).
```

### CODE REVIEW

**Comment by dkorolev** [comment]
X1:

```
N = 100, M = 1
Memory
Iteration 1/10, 3644.932072 replications per second.
Iteration 2/10, 11318.166228 replications per second.
Iteration 3/10, 12348.758887 replications per second.
Iteration 4/10, 11376.695013 replications per second.
Iteration 5/10, 11141.120864 replications per second.
Iteration 6/10, 11637.475021 replications per second.
Iteration 7/10, 13945.626145 replications per second.
Iteration 8/10, 8379.179010 replications per second.
Iteration 9/10, 11743.772242 replications per second.
Iteration 10/10, 12708.600770 replications per second.
Replications per second: 10824.432625 (stddev 2743.998613).
File
Iteration 1/10, 12530.059486 replications per second.
Iteration 2/10, 8189.940437 replications per second.
Iteration 3/10, 15228.426396 replications per second.
Iteration 4/10, 9823.377654 replications per second.
Iteration 5/10, 9134.526665 replications per second.
Iteration 6/10, 10129.949862 replications per second.
Iteration 7/10, 10723.570191 replications per second.
Iteration 8/10, 11729.857820 replications per second.
Iteration 9/10, 11573.532850 replications per second.
Iteration 10/10, 12318.029115 replications per second.
Replications per second: 11138.127048 (stddev 1903.233576).
```

**Comment by dkorolev** [comment]
Hetzner:

```
N = 100, M = 1
Memory
Iteration 1/10, 19031.141869 replications per second.
Iteration 2/10, 13736.644929 replications per second.
Iteration 3/10, 14920.874152 replications per second.
Iteration 4/10, 15235.457064 replications per second.
Iteration 5/10, 9205.876883 replications per second.
Iteration 6/10, 15384.615385 replications per second.
Iteration 7/10, 9580.027095 replications per second.
Iteration 8/10, 8144.796380 replications per second.
Iteration 9/10, 13299.301451 replications per second.
Iteration 10/10, 30697.674419 replications per second.
Replications per second: 14923.640963 (stddev 6142.056961).
File
Iteration 1/10, 19635.065450 replications per second.
Iteration 2/10, 9755.616870 replications per second.
Iteration 3/10, 10440.835267 replications per second.
Iteration 4/10, 9560.598745 replications per second.
Iteration 5/10, 11003.667889 replications per second.
Iteration 6/10, 11322.049405 replications per second.
Iteration 7/10, 5914.685148 replications per second.
Iteration 8/10, 7284.232212 replications per second.
Iteration 9/10, 10618.899496 replications per second.
Iteration 10/10, 11042.944785 replications per second.
Replications per second: 10657.859527 (stddev 3424.105042).
```

**Comment by dkorolev** [comment]
Large log size slows down both memory and disk persisters, hinting it's Sherlock (as I doubt a few dozen megabytes of JSON parsing could be that bad).

**Comment by mzhurovich** [comment]
@dkorolev Let's move it to `examples/Benchmark`?

**Comment by dkorolev** [comment]
@mzhurovich perhaps into `regression_tests/replication` then?

**Comment by dkorolev** [comment]
OK, after the bugfix, this test does OK:

```
./.current/replications_per_second -n 4 -m 9999999
while true ; do (curl -s localhost:8500/stream?sizeonly; curl -s localhost:8501/stream?sizeonly ; curl -s localhost:8502/stream?sizeonly ; curl -s localhost:8503/stream?sizeonly) | paste - - - - ; sleep 1 ; done
```

And `./.current/replications_per_second -n 100` is now higher "replications per second" than `./.current/replications_per_second -n 10` on Hetzner.

Looks like the replication itself (one leg) is slow, while chaining doesn't add much to the end-to-end performance (the delay doesn't really accumulate leg-to-leg).

**Comment by dkorolev** [comment]
*[ Testing on our Hetzner machine ]*

Replicating a lot of data over two legs, Follower 2 <- Follower 1 <- Master.

```
[ terminal 1 ] $ ./.current/replications_per_second  -n 3 -m 10000000

[ terminal 2 ] $ while true ; do (curl -s localhost:8502/stream?sizeonly; curl -s localhost:8501/stream?sizeonly ; curl -s localhost:8500/stream?sizeonly) | paste - - - ; sleep 1 ; done

...

293154  293168  10000000
295068  295082  10000000
296998  297012  10000000
```

=> **The lag is effectively zero (`curl` time eats up more), and the replication speed is a but under 2K records per second.**

Now, change two legs into 99 legs (`-n 100 -m 10000000`), and change `localhost:8592` into `localhost:8599` in the second command:

```
107987  108003  10000000
108581  108599  10000000
109182  109193  10000000
```

=> **Same picture, thus adding more legs doesn't make the end-to-end replication slower, thus the lag is indeed negligible.**


We have confirmed:

1. The replication itself is stable on localhost (although ports do get eaten up as we run more "parallelized" tests. -- D.K.), and

2. The lag introduced by replication is sub-sub second (it's within 10% range comparing replicating over two legs vs. replicating over a hundred legs).

File: examples/Benchmark/replication/replications_per_second.cc [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    replicators[i] = std::make_unique<current::sherlock::StreamReplicator<stream_t>>(*streams[i]);
+  }
+
+  // Start the chain of replications.
+  std::vector<current::sherlock::SubscriberScope> subscriber_scopes(FLAGS_n);
+  for (uint32_t i = 1; i < FLAGS_n; ++i) {
+    subscriber_scopes[i] = remote_subscribers[i - 1]->Subscribe(*replicators[i]);
+  }
+
+  // Publish M events and watch them proparate.
```
Comment (mzhurovich) [inline examples/Benchmark/replication/replications_per_second.cc]: s/proparate/propagate/

File: examples/Benchmark/replication/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+```
+...
+
+293154  293168  10000000
+295068  295082  10000000
+296998  297012  10000000
+```
+
+=> **The lag is effectively zero (`curl` time eats up more), and the replication speed is a but under 2K records per second.**
```
Comment (mzhurovich) [inline examples/Benchmark/replication/README.md]: s/but/bit/

**Comment by dkorolev** [comment]
@mzhurovich, PTAL.

**Pull request merged**